src/script/interpreter.cpp

```diff
@@ -2092,6 +2092,92 @@ bool VerifyTaprootCommitment(uint256 &tapleaf_hash,
+            valtype(script_pubkey.begin() + SCRIPT_INTRO_SIZE,
+                    script_pubkey.begin() + SCRIPT_SIZE_WITHOUT_STATE);
+
+        if (stack.size() == 0) {
+            return set_error(serror, ScriptError::INVALID_STACK_OPERATION);
+        }
+        if (stack.size() >= 2 && !stack.back().empty() &&
+            stack.back()[0] == ANNEX_TAG) {
+            return set_error(serror, ScriptError::TAPROOT_ANNEX_NOT_SUPPORTED);
+        }
+        if (stack.size() == 1) {
+            // Spend using single signature instead of executing script
+            valtype &vch_sig = stacktop(-1);
+            bool f_success;
+            if (!EvalChecksig(vch_sig, vch_pubkey, script_pubkey.begin(),
+                              script_pubkey.end(),
+                              flags | SCRIPT_REQUIRE_TAPROOT_SIGHASH, checker,
+                              metrics_out, std::nullopt, serror, f_success)) {
+                // serror is set
+                return false;
+            }
+            if (!f_success) {
+                return set_error(serror,
+                                 ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED);
+            }
+            return set_success(serror);
+        } else {
+            // Spend using executing script, internal pubkey and merkle path
+            valtype control_block = stacktop(-1);
+            valtype script_bytes = stacktop(-2);
+            CScript exec_script(script_bytes.begin(), script_bytes.end());
+            popstack(stack);
+            popstack(stack);
+            const uint32_t size_remainder =
+                (control_block.size() - TAPROOT_CONTROL_BASE_SIZE) %
+                TAPROOT_CONTROL_NODE_SIZE;
+            if (control_block.size() < TAPROOT_CONTROL_BASE_SIZE ||
+                control_block.size() > TAPROOT_CONTROL_MAX_SIZE ||
+                (size_remainder != 0)) {
+                return set_error(serror,
+                                 ScriptError::TAPROOT_WRONG_CONTROL_SIZE);
+            }
+            if ((control_block[0] & TAPROOT_LEAF_MASK) !=
+                TAPROOT_LEAF_TAPSCRIPT) {
+                return set_error(
+                    serror, ScriptError::TAPROOT_LEAF_VERSION_NOT_SUPPORTED);
+            }
+            uint256 tapleaf_hash;
+            if (!VerifyTaprootCommitment(tapleaf_hash, control_block,
+                                         vch_pubkey, exec_script)) {
+                return set_error(serror,
+                                 ScriptError::TAPROOT_VERIFY_COMMITMENT_FAILED);
+            }
+            if (script_pubkey.size() == SCRIPT_SIZE_WITH_STATE) {
+                stack.push_back(valtype(
+                    script_pubkey.begin() + SCRIPT_SIZE_WITHOUT_STATE + 1,
+                    script_pubkey.begin() + SCRIPT_SIZE_WITH_STATE));
+            }
+            ScriptExecutionData execdata{tapleaf_hash};
+            if (!EvalScript(stack, exec_script, flags, checker, metrics_out,
```

Comment:
Don't you need `flags | SCRIPT_REQUIRE_TAPROOT_SIGHASH` here?

Reply:
Given these new sighash types they kind of modify the behavior of the sighashes?
In the case of BIP341 with P2SH the scriptCode would be the `redeemScript`.
In the case of BIP341 with TAPROOT exec path, the scriptCode would be the `redeemScript`.
In the case of BIP341 with TAPROOT sign path, the exec data is not used.

In the case of BIP143 with taproot, the scriptcode would be the `tapleaf_hash` (NOT the actual code).
In the case of BIP143 with p2sh without taproot the scriptcode would be the redeemScript
In the case of BIP143 with p2pkh the scriptCode would be the pubKey(?) (or the pubkey hash?)

And then all the legacy hash types ... I don't care to figure this out right now.

Reply:
Correction:

1. In the case of BIP341 with P2SH the `executed_script_hash` would be the `SHA256(redeemScript)`.
2. correct
3. correct
4. In the case of BIP143 with taproot, the scriptCode would be the `exec_script` from the last executed OP_CODESEPARATOR.
5. In the case of BIP143 with p2sh without taproot the scriptcode would be the `redeemScript` from the last executed OP_CODESEPARATOR
6. In the case of BIP143 with p2pkh the scriptCode would be the scriptPubKey